File: internal/models/workflow.go
package models

import (
	"database/sql/driver"
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

// Workflow represents a workflow definition
type Workflow struct {
	ID             uuid.UUID          `json:"id" db:"id"`
	OrganizationID uuid.UUID          `json:"organization_id" db:"organization_id"`
	WorkflowID     string             `json:"workflow_id" db:"workflow_id"`
	Version        string             `json:"version" db:"version"`
	Name           string             `json:"name" db:"name"`
	Description    *string            `json:"description,omitempty" db:"description"`
	Definition     WorkflowDefinition `json:"definition" db:"definition"`
	Enabled        bool               `json:"enabled" db:"enabled"`
	CreatedAt      time.Time          `json:"created_at" db:"created_at"`
	UpdatedAt      time.Time          `json:"updated_at" db:"updated_at"`
	CreatedBy      *uuid.UUID         `json:"created_by,omitempty" db:"created_by"`
	Tags           []string           `json:"tags,omitempty" db:"tags"`
}

// WorkflowDefinition represents the complete workflow definition
type WorkflowDefinition struct {
	Trigger TriggerDefinition `json:"trigger"`
	Context ContextDefinition `json:"context,omitempty"`
	Steps   []Step            `json:"steps"`
	Timeout string            `json:"timeout,omitempty"` // Global timeout duration, e.g., "5m", "1h", "30s"
}

// TriggerDefinition defines what starts the workflow
type TriggerDefinition struct {
	Type  string                 `json:"type"` // event, schedule, manual
	Event string                 `json:"event,omitempty"`
	Cron  string                 `json:"cron,omitempty"`
	Data  map[string]interface{} `json:"data,omitempty"`
}

// ContextDefinition defines what data to load
type ContextDefinition struct {
	Load []string `json:"load,omitempty"` // e.g., ["order.details", "customer.history"]
}

// Step represents a workflow step
type Step struct {
	ID        string                 `json:"id"`
	Type      string                 `json:"type"` // condition, action, parallel, execute, wait
	Condition *Condition             `json:"condition,omitempty"`
	Action    *Action                `json:"action,omitempty"`
	OnTrue    string                 `json:"on_true,omitempty"`
	OnFalse   string                 `json:"on_false,omitempty"`
	Parallel  *ParallelStep          `json:"parallel,omitempty"`
	Execute   []ExecuteAction        `json:"execute,omitempty"`
	Wait      *WaitConfig            `json:"wait,omitempty"`
	Retry     *RetryConfig           `json:"retry,omitempty"`
	Timeout   string                 `json:"timeout,omitempty"` // Step-level timeout, e.g., "30s", "2m"
	Metadata  map[string]interface{} `json:"metadata,omitempty"`
}

// Condition represents a conditional expression
type Condition struct {
	Field    string      `json:"field"`
	Operator string      `json:"operator"` // eq, neq, gt, gte, lt, lte, in, contains, regex
	Value    interface{} `json:"value"`
	And      []Condition `json:"and,omitempty"`
	Or       []Condition `json:"or,omitempty"`
}

// Action represents an action to take
type Action struct {
	Type     string                 `json:"action"` // allow, block, execute
	Reason   string                 `json:"reason,omitempty"`
	Metadata map[string]interface{} `json:"metadata,omitempty"`
}

// ParallelStep represents parallel execution of steps
type ParallelStep struct {
	Steps    []Step `json:"steps"`
	Strategy string `json:"strategy"` // all_must_pass, any_can_pass, best_effort
}

// ExecuteAction represents an action to execute
type ExecuteAction struct {
	Type       string                 `json:"type"` // notify, webhook, create_record, http_request, etc.
	Recipients []string               `json:"recipients,omitempty"`
	Message    string                 `json:"message,omitempty"`
	URL        string                 `json:"url,omitempty"`
	Method     string                 `json:"method,omitempty"`
	Headers    map[string]string      `json:"headers,omitempty"`
	Body       map[string]interface{} `json:"body,omitempty"`
	Entity     string                 `json:"entity,omitempty"`
	EntityID   string                 `json:"entity_id,omitempty"`
	Data       map[string]interface{} `json:"data,omitempty"`
}

// WaitConfig represents wait/timeout configuration
type WaitConfig struct {
	Event     string `json:"event"`
	Timeout   string `json:"timeout"` // duration string, e.g., "24h"
	OnTimeout string `json:"on_timeout"`
}

// RetryConfig represents retry configuration
type RetryConfig struct {
	MaxAttempts int      `json:"max_attempts"`
	Backoff     string   `json:"backoff"` // linear, exponential
	RetryOn     []string `json:"retry_on,omitempty"`
}

// CreateWorkflowRequest represents the request to create a workflow
type CreateWorkflowRequest struct {
	WorkflowID  string             `json:"workflow_id" validate:"required"`
	Version     string             `json:"version" validate:"required"`
	Name        string             `json:"name" validate:"required"`
	Description *string            `json:"description,omitempty"`
	Definition  WorkflowDefinition `json:"definition" validate:"required"`
	Tags        []string           `json:"tags,omitempty"`
}

// UpdateWorkflowRequest represents the request to update a workflow
type UpdateWorkflowRequest struct {
	Name        *string             `json:"name,omitempty"`
	Description *string             `json:"description,omitempty"`
	Definition  *WorkflowDefinition `json:"definition,omitempty"`
	Tags        []string            `json:"tags,omitempty"`
}

// JSONB scanning for WorkflowDefinition
func (w *WorkflowDefinition) Scan(value interface{}) error {
	if value == nil {
		return nil
	}

	bytes, ok := value.([]byte)
	if !ok {
		return nil
	}

	return json.Unmarshal(bytes, w)
}

func (w WorkflowDefinition) Value() (driver.Value, error) {
	return json.Marshal(w)
}

File: web/src/types/workflow.ts
export interface Workflow {
  id: string
  workflow_id: string
  version: string
  name: string
  description?: string
  definition: WorkflowDefinition
  enabled: boolean
  created_at: string
  updated_at: string
  tags?: string[]
}

export interface WorkflowDefinition {
  workflow_id: string
  version: string
  name: string
  description?: string
  enabled: boolean
  trigger: Trigger
  context?: ContextConfig
  steps: Step[]
}

export interface Trigger {
  type: 'event' | 'schedule' | 'manual'
  event?: string
  schedule?: string
}

export interface ContextConfig {
  load?: string[]
}

export interface Step {
  id: string
  type: 'condition' | 'action' | 'execute' | 'parallel' | 'foreach'
  name?: string
  condition?: Condition
  action?: Action
  execute?: ExecuteAction[]
  parallel?: ParallelStep
  on_true?: string
  on_false?: string
  next?: string
  metadata?: Record<string, any>
  retry?: RetryConfig
  // Legacy fields for backwards compatibility
  steps?: Step[]
  strategy?: string
}

export interface Condition {
  field?: string
  operator?: 'eq' | 'neq' | 'gt' | 'gte' | 'lt' | 'lte' | 'in' | 'contains'
  value?: any
  and?: Condition[]
  or?: Condition[]
  not?: Condition
}

export interface Action {
  type: 'allow' | 'block' | 'execute' | 'wait' | 'require_approval'
  reason?: string
}

export interface ExecuteAction {
  type: 'notify' | 'webhook' | 'http_request' | 'create_record' | 'update_record' | 'create_approval_request' | 'log'
  // Notify action fields
  recipients?: string[]
  message?: string
  // Webhook/HTTP request fields
  url?: string
  method?: string
  headers?: Record<string, string>
  body?: Record<string, any>
  // Record action fields
  entity?: string
  entity_id?: string
  data?: Record<string, any>
}

export interface ParallelStep {
  steps: Step[]
  strategy: 'all_must_pass' | 'any_can_pass' | 'best_effort'
}

export interface RetryConfig {
  max_attempts?: number
  backoff?: string
  retry_on?: string[]
}

export interface Execution {
  id: string
  execution_id: string
  workflow_id: string
  trigger_event: string
  trigger_payload: Record<string, any>
  context?: Record<string, any>
  status: ExecutionStatus
  result?: ExecutionResult
  started_at: string
  completed_at?: string
  duration_ms?: number
  error_message?: string
  steps?: StepExecution[]
}

export type ExecutionStatus = 'pending' | 'running' | 'waiting' | 'completed' | 'failed' | 'blocked' | 'cancelled' | 'paused'
export type ExecutionResult = 'allowed' | 'blocked' | 'executed'

export interface StepExecution {
  id: string
  execution_id: string
  step_id: string
  step_type: string
  status: ExecutionStatus
  input?: Record<string, any>
  output?: Record<string, any>
  result?: any
  started_at: string
  completed_at?: string
  duration_ms?: number
  error_message?: string
}

export interface ExecutionTraceResponse {
  execution: Execution
  steps: StepExecution[]
  workflow?: Workflow
}

export interface ApprovalRequest {
  id: string
  request_id: string
  execution_id: string
  entity_type: string
  entity_id: string
  requester_id: string
  approver_role: string
  approver_id?: string
  status: ApprovalStatus
  reason: string
  decision_reason?: string
  requested_at: string
  decided_at?: string
  expires_at?: string
}

export type ApprovalStatus = 'pending' | 'approved' | 'rejected' | 'expired'

export interface CreateWorkflowRequest {
  workflow_id: string
  version: string
  name: string
  description?: string
  definition: WorkflowDefinition
  tags?: string[]
}

export interface UpdateWorkflowRequest {
  name?: string
  description?: string
  definition?: WorkflowDefinition
  tags?: string[]
}

export interface IngestEventRequest {
  event_type: string
  source: string
  payload: Record<string, any>
}
